go/store/prolly/tree/merge.go

```diff
@@ -88,172 +81,295 @@ func ThreeWayMerge[K ~[]byte, O Ordering[K], S message.Serializer](
+// usually this is the level of the |to| cursor, but if that cursor is exhausted,
+// then the patch generator is emitting removed diffs on the level of the |from| cursor.
+func (d *PatchGenerator[K, O]) getLevel() (uint64, error) {
+	if d.to.Valid() {
+		return d.to.level()
 	}
+	return d.from.level()
 }
 
-func (ps PatchBuffer) Close() error {
-	close(ps.buf)
-	return nil
+// resolveCollision takes two point Patches with the same key and resolves the changes.
+func resolveCollision(left Patch, lDiffType DiffType, right Patch, rDiffType DiffType, cb CollisionFn) (merged Patch, ok bool) {
+	leftDiff := Diff{
+		Key:  left.EndKey,
+		From: left.From,
+		To:   left.To,
+		Type: lDiffType,
+	}
+	rightDiff := Diff{
+		Key:  right.EndKey,
+		From: right.From,
+		To:   right.To,
+		Type: rDiffType,
+	}
+	resolved, ok := cb(leftDiff, rightDiff)
+	return Patch{
+		From:   resolved.From,
+		EndKey: resolved.Key,
+		To:     resolved.To,
+		Level:  0,
+	}, ok
 }
 
-func sendPatches[K ~[]byte, O Ordering[K]](
+// SendPatches iterates over |l| and |r| in parallel, sending an ordered non-overlapping series of patches into |buf|.
+func SendPatches[K ~[]byte, O Ordering[K]](
 	ctx context.Context,
-	l, r Differ[K, O],
+	l, r PatchGenerator[K, O],
 	buf PatchBuffer,
 	cb CollisionFn,
-) (stats MergeStats, err error) {
+) (err error) {
 	var (
-		left, right Diff
-		lok, rok    = true, true
+		left, right          Patch
+		lDiffType, rDiffType DiffType
+		lok, rok             = true, true
 	)
 
-	left, err = l.Next(ctx)
+	left, lDiffType, err = l.Next(ctx)
 	if err == io.EOF {
 		err, lok = nil, false
 	}
 	if err != nil {
-		return MergeStats{}, err
+		return err
 	}
 
-	right, err = r.Next(ctx)
+	right, rDiffType, err = r.Next(ctx)
 	if err == io.EOF {
 		err, rok = nil, false
 	}
 	if err != nil {
-		return MergeStats{}, err
+		return err
 	}
 
 	for lok && rok {
-		cmp := l.order.Compare(ctx, K(left.Key), K(right.Key))
+		order := l.order
+		// If they're ranges, compare the start points, see if they overlap.
+		leftLevel, _ := l.getLevel()
+		rightLevel, _ := r.getLevel()
+		if leftLevel > 0 && rightLevel > 0 {
+			if nilCompare(ctx, order, K(left.EndKey), K(right.KeyBelowStart)) <= 0 {
+				// Left change is entirely before right change.
+				// This change is already on the left map, so we ignore it.
+				left, lDiffType, err = l.Next(ctx)
+				if err == io.EOF {
+					err, lok = nil, false
+				}
+				if err != nil {
+					return err
+				}
+			} else if nilCompare(ctx, order, K(right.EndKey), K(left.KeyBelowStart)) <= 0 {
+				// Right change is entirely before right change.
+				err = buf.SendPatch(ctx, right)
+				if err != nil {
+					return err
+				}
+
+				right, rDiffType, err = r.Next(ctx)
+				if err == io.EOF {
+					err, rok = nil, false
+				}
+				if err != nil {
+					return err
+				}
+			} else if bytes.Equal(left.To, right.To) {
+				// A concurrent change.
```

Comment:
This is a bit subtle. Might be worth commenting that `EndKey` is fully determined by `To` here because levels > 0, so that these two patches are pointing to identical contents of within their merkle trees.